### PR TITLE
Typed Quotes

### PR DESCRIPTION
**Pull request by bloderxd**
# What
This PR brings the implementation of [Typed Quotes](https://github.com/arrow-kt/arrow-meta/issues/673), with it we bring more type information for quotes being possible solutions involving subtyping or any other kind related with descriptors.

# How
All Typed Quotes should rewind the analysis phase binding analysed descriptors with its related elements but Kotlin compiler only allows this rewind only one time, this is not good since we can have N plugins that have one or more Typed Quotes rewinding analysis phase. Thinking about that `Analysis Context` was created to represent and control all Arrow Meta plugins and its rewinding behavior.

# Typed Quotes implemented
This PR is migrating 3 Quotes to Typed Quotes: `NamedFunction`, `Property` and `TypeAlias`

# Next steps
I see two important next steps:

- Other important quotes should be migrated to Typed Quotes
- Study providing more information for Typed Quotes such resolved calls

### CODE REVIEW

File: compiler-plugin/src/main/kotlin/arrow/meta/plugins/proofs/phases/quotes/SerializeRefinedPredicates.kt [hunk 1]
```diff
@@ -1,19 +1,28 @@
...
 import arrow.meta.phases.analysis.ElementScope
+import arrow.meta.phases.evaluateDependsOn
 import arrow.meta.quotes.Transform
 import arrow.meta.quotes.classorobject.ObjectDeclaration
 import org.jetbrains.kotlin.psi.KtObjectDeclaration
 
-internal fun ObjectDeclaration.objectWithSerializedRefinement(elementScope: ElementScope): Transform<KtObjectDeclaration> =
-  elementScope.run {
+internal fun ObjectDeclaration.objectWithSerializedRefinement(elementScope: ElementScope, ctx: CompilerContext): Transform<KtObjectDeclaration> =
+  ctx.evaluateDependsOn(
```
Comment (raulraja) [inline compiler-plugin/src/main/kotlin/arrow/meta/plugins/proofs/phases/quotes/SerializeRefinedPredicates.kt:11]: Is there a way to encapsulate this reoccurring pattern in the internals so it does not have to be repeated in each quote? Seems boilerplate except `evaluatesRefinementExpression(elementScope)`

**Review by rachelcarmena — APPROVED** [review]
Awesome @bloderxd :tada: Please, take into account [this change](https://github.com/arrow-kt/arrow-meta/pull/752) after merging `master` branch.

**Comment by raulraja** [comment]
@bloderxd any chance we can get this fixed and merged? Some users are asking for this feature 👌

**Comment by rachelcarmena** [comment]
Thanks @bloderxd !! I'll fix the new failures from checks in another PR before :raised_hands:

**Comment by raulraja** [comment]
Thanks @bloderxd !

**Pull request merged**